### PR TITLE
Add ApiEndpoint support to GCE config.

### PR DESCRIPTION
**Pull request by cadmuxe**
**What this PR does / why we need it**:
Add the ability to change ApiEndpoint  for GCE.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:
```release-note
None
```

### CODE REVIEW

**Comment by cadmuxe** [comment]
CLA Signed.

**Comment by freehan** [comment]
@k8s-bot ok to test

**Comment by freehan** [comment]
What about exposing an environment variable in kube-up.sh? Or do it in a follow up PR.

This way, one can just specified `KUBE_GCP_API_ENDPOINT=foo` to bring up a custom cluster.

**Comment by freehan** [comment]
FYI, it seems like test/e2e/e2e.go is also referencing this.

**Comment by cadmuxe** [comment]
Yes, you are correct. I will update this PR.
I had an change for test/e2e/e2e.go,  please check this commit
https://github.com/kubernetes/kubernetes/pull/47881/commits/890d356d5c788f85aa542ac945c12fc78fb8846c

Thanks

**Comment by cadmuxe** [comment]
@freehan please check this commit for the environment variable change. Thanks.
https://github.com/kubernetes/kubernetes/pull/47881/commits/b3956a689eaa7730e3fe3c1ca253a4027f9a1b9f

**Comment by freehan** [comment]
Cool! Thanks! 
/lgtm
@fejta, @bowei for /approve

File: cluster/gce/config-default.sh [hunk 1]
```diff
@@ -72,6 +72,7 @@ GCI_VERSION=${KUBE_GCI_VERSION:-cos-stable-59-9460-64-0}
 MASTER_IMAGE=${KUBE_GCE_MASTER_IMAGE:-}
 MASTER_IMAGE_PROJECT=${KUBE_GCE_MASTER_PROJECT:-cos-cloud}
 NODE_IMAGE=${KUBE_GCE_NODE_IMAGE:-${CVM_VERSION}}
+KUBE_GCE_API_ENDPOINT=${KUBE_GCE_API_ENDPOINT:-}
```
Comment (freehan) [inline cluster/gce/config-default.sh]: nit: rename the left var to `GCE_API_ENDPOINT`

File: cluster/gce/gci/configure-helper.sh [hunk 2]
```diff
@@ -379,6 +379,11 @@ function create-master-auth {
   cat <<EOF >/etc/gce.conf
 [global]
 EOF
+  if [[ -n "${KUBE_GCE_API_ENDPOINT:-}" ]]; then
```
Comment (freehan) [inline cluster/gce/gci/configure-helper.sh]: nit: use `GCE_API_ENDPOINT`

File: cluster/gce/gci/configure-helper.sh [hunk 3]
```diff
@@ -379,6 +379,11 @@ function create-master-auth {
...
+    cat <<EOF >>/etc/gce.conf
+api-endpoint = ${KUBE_GCE_API_ENDPOINT}
```
Comment (freehan) [inline cluster/gce/gci/configure-helper.sh]: same here

**Review by freehan — COMMENTED** [review]
Sorry I take lgtm back

File: cluster/common.sh [hunk 4]
```diff
@@ -665,6 +665,7 @@ ENABLE_APISERVER_BASIC_AUDIT: $(yaml-quote ${ENABLE_APISERVER_BASIC_AUDIT:-})
 ENABLE_APISERVER_ADVANCED_AUDIT: $(yaml-quote ${ENABLE_APISERVER_ADVANCED_AUDIT:-})
 ENABLE_CACHE_MUTATION_DETECTOR: $(yaml-quote ${ENABLE_CACHE_MUTATION_DETECTOR:-false})
 ADVANCED_AUDIT_BACKEND: $(yaml-quote ${ADVANCED_AUDIT_BACKEND:-log})
+KUBE_GCE_API_ENDPOINT: $(yaml-quote ${KUBE_GCE_API_ENDPOINT:-})
```
Comment (freehan) [inline cluster/common.sh]: why do you need this?

Comment (freehan) [inline cluster/common.sh]: same here

**Comment by freehan** [comment]
`hack/verify-gofmt.sh` and `hack/verify-flags-underscore.py` failed. Please try to run them and fix the errors.

File: cluster/common.sh (hunk 4 above)
Comment (cadmuxe) [inline cluster/common.sh]: Done, Thanks.

File: cluster/gce/config-default.sh (hunk 1 above)
Comment (cadmuxe) [inline cluster/gce/config-default.sh]: Done, Thanks.

File: cluster/gce/gci/configure-helper.sh (hunk 2 above)
Comment (cadmuxe) [inline cluster/gce/gci/configure-helper.sh]: Done, Thanks.

File: cluster/gce/gci/configure-helper.sh (hunk 3 above)
Comment (cadmuxe) [inline cluster/gce/gci/configure-helper.sh]: Done, Thanks.

**Comment by nicksardo** [comment]
@cadmuxe There are several areas in the cloud provider which still hardcode the api endpoint.  I can fix these in a PR which I'm currently working on.

There's another issue which I'm not sure how to answer: the package also consumes the beta client package: https://github.com/kubernetes/kubernetes/blob/7df2bce1ece551800edb1a8f1f62adecbd11cff5/pkg/cloudprovider/providers/gce/gce.go#L256; however, your PR doesn't change the base path for that. Its default value is `"https://www.googleapis.com/compute/beta/projects/"`

**Comment by cadmuxe** [comment]
@nicksardo nope, I forgot to update the endpoint for the beta client. Cloud you also fix this in your PR? Thanks

**Pull request merged**